Enable TLS

ishustava opened the pull request:
This PR builds on the work done in #137 and proposes the following changes:

* Optionally allows you to turn on TLS by setting `global.tls.enabled` to `true`.
* Exposes the default HTTPS port 8501 on clients and servers if TLS enabled.
* Optionally allows you to disable HTTP ports on clients and servers by setting `global.tls.httpsOnly` to `true`. This is set to `true` by default.
* Adds TLS bootstrapping job that generates Consul cluster CA and server certificate.
* Adds `tls-init-cleanup` job that deletes certs created by the `tls-init` job when the Helm chart is deleted.
* Consul client certs are generated in an init container, so that we can add `HOST_IP` as a SAN to enable other clients, e.g. sync-catalog, to talk to it. This will be replaced by `auto_encrypt` in the future once it supports providing additional SANs to the client certificate.
* Doesn't expose `verify_*` config options in the Helm chart as it was proposed in the original PR #137. Instead, it assumes secure defaults if `global.tls.enabled` is `true`.

Note 1: This currently doesn't support external CAs. It also doesn't work if servers aren't deployed on k8s. Once we support external CAs, we can also support various deployment configurations. Hence, if servers are disabled, `tls-init` will not run.

Note 2: We're not using mTLS for HTTP API. That is because in [Consul's threat model](https://www.consul.io/docs/internals/security.html#secure-configuration) we recommend using ACLs instead. To prevent clients from sending unencrypted RPC requests to the servers, we set `verify_incoming_rpc` to `true`.

Fixes #136

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -71,6 +71,14 @@ spec:
+            secretName: {{ template "consul.fullname" . }}-ca-cert
+        - name: tls-client-cert
+          emptyDir:
+            medium: "Memory"
```
Using `tmpfs` to store client certs.

lkysow replied:
fancy! Do you know which versions of kube that's supported with?

ishustava replied:
good question! It's supported in 1.13+, don't know about earlier versions.

lkysow commented on values.yaml:
```diff
@@ -52,6 +52,35 @@ global:
 
+  # Enables TLS encryption across the cluster to verify authenticity of the
+  # servers and clients that connect. Note: It is HIGHLY recommended that you also
+  # enable Gossip encryption.
```
This should really link to our own docs page on enabling gossip encryption... if we had one.

lkysow commented on templates/tls-init-job.yaml:
```diff
@@ -0,0 +1,65 @@
+                -H "Authorization: Bearer $( cat /var/run/secrets/kubernetes.io/serviceaccount/token )" \
+                -H "Content-Type: application/json" \
+                -H "Accept: application/json" \
+                -d "{ \"kind\": \"Secret\", \"apiVersion\": \"v1\", \"metadata\": { \"name\": \"{{ template "consul.fullname" . }}-server-cert\", \"namespace\": \"${NAMESPACE}\" }, \"type\": \"kubernetes.io/tls\", \"data\": { \"tls.crt\": \"$( cat {{ .Values.global.datacenter }}-server-{{ .Values.global.domain }}-0.pem | base64 | tr -d '\n' )\", \"tls.key\": \"$( cat {{ .Values.global.datacenter  ...
```
There's one cert for all the servers right? So isn't this weird it's suffixed with -0?

ishustava replied:
That is what consul CLI does by default. It doesn't matter so much though because we're uploading only the contents here. 

It's maybe weird for client certs. Each client cert has the -0 suffix, and so it could be a bit confusing. We could rename them, but with auto_encrypt we won't need this anymore.

lkysow commented on templates/tls-init-cleanup-serviceaccount.yaml:
```diff
@@ -0,0 +1,16 @@
+    app: {{ template "consul.name" . }}
+    chart: {{ template "consul.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
```
missing `annotations:`

ishustava replied:
Good catch! That's funny, it looks like they are not needed here actually. So will remove them altogether.

lkysow commented on templates/tls-init-cleanup-clusterrole.yaml:
```diff
@@ -0,0 +1,25 @@
+    chart: {{ template "consul.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  "helm.sh/hook": post-delete
```
need annotations:

lkysow commented on templates/tls-init-cleanup-clusterrolebinding.yaml:
```diff
@@ -0,0 +1,24 @@
+    chart: {{ template "consul.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  "helm.sh/hook": post-delete
```
annotations:

ishustava commented:
This PR requires `connectInject` to support TLS before merging.

lkysow commented on test/unit/client-daemonset.bats:
```diff
@@ -418,9 +418,205 @@ load _helpers
+#--------------------------------------------------------------------
+# global.tls.enabled
+
+@test "client/DaemonSet: no secret volumes when TLS is disabled" {
```
seems fragile, what if I add a secret for something unrelated? can you test just that it doesn't have the tls volumes? Or... delete this test since checking the absence of things for each feature seems like overkill

ishustava replied:
Yeah I think you're right. Deleted the test.

lkysow commented on test/unit/client-daemonset.bats:
```diff
@@ -418,9 +418,205 @@ load _helpers
+  [ "${actual}" != "" ]
+}
+
+@test "client/DaemonSet: no volumes mounted when TLS is disabled" {
```
see same comment above

ishustava replied:
Deleted this test.

lkysow commented on test/unit/client-daemonset.bats:
```diff
@@ -418,9 +418,205 @@ load _helpers
+  [ "${actual}" != "" ]
+}
+
+@test "client/DaemonSet: port 8500 is exposed when TLS is disabled" {
```
this should just be "by default"

ishustava replied:
Yep, deleted this one too.

lkysow commented on test/unit/client-daemonset.bats:
```diff
@@ -418,9 +418,205 @@ load _helpers
+      . | tee /dev/stderr |
+      yq '.spec.template.spec.initContainers[] | select(.name == "client-tls-init")' | tee /dev/stderr)
+
+  # test that it generates client certificate
```
seems like overkill, e.g. why not test the exact command? i think testing the init container exists is enough

ishustava replied:
Hmm yeah, I suppose that's true. I've removed these assertions and replaced them with an assertion that the init container exists, as you suggested.

lkysow commented on values.yaml:
```diff
@@ -52,6 +52,35 @@ global:
+  tls:
+    enabled: false
+
+    # Set serverAdditionalDNSSANs is a list of additional DNS names to
```
remove the word `Set` from these

ishustava replied:
Done!

lkysow commented on values.yaml:
```diff
@@ -52,6 +52,35 @@ global:
+    # for example, if you're using the UI.
+    serverAdditionalIPSANs: []
+
+    # If httpsOnly is true, Consul will disable the HTTP port and only accept
```
not clear if this is across server, clients or both

ishustava replied:
Updated!

lkysow commented on values.yaml:
```diff
@@ -52,6 +52,35 @@ global:
+  #
+  # Note: this relies on functionality introduced with Consul 1.4.1. Make sure
+  # your global.image value is at least version 1.4.1.
+  tls:
```
https://www.youtube.com/watch?v=dAVyEFK1Cs4&t=33s

ishustava replied:
🤣

lkysow commented:
Outstanding question from me: How do users rotate their certificates?

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -126,8 +133,20 @@ spec:
                 -advertise="${ADVERTISE_IP}" \
                 -bind=0.0.0.0 \
                 -client=0.0.0.0 \
+                {{- if .Values.global.tls.enabled }}
```
Yeah, I think it would be nice. My only reservation against it is that it currently requires a manual restart. So disabling/enabling through the Helm chart on an existing cluster won't do anything.

lkysow replied:
Oh good catch. Okay.

lkysow commented on templates/tls-init-job.yaml:
```diff
@@ -0,0 +1,65 @@
+                 {{- end }}
+                -dc={{ .Values.global.datacenter }} \
+                -domain={{ .Values.global.domain }}
+              curl -s -X POST --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
```
does it ever make sense for the CA to be a separate secret from the key? So I can give access to the CA for validation without also giving the key, i.e. in rbac `resourceNames`

ishustava replied:
Yeah, I think it does. I'll split them up into two separate secrets.

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -177,9 +205,16 @@ spec:
                 - -c
                 - consul leave
```
I'm so glad you asked this! After some testing, I've found that consul clients already have `leave_on_terminate` set to true by default (as of consul 0.7) according to the [docs](https://www.consul.io/docs/agent/options.html#leave_on_terminate). I'll set that value to true for clients, so that it works even for an earlier version. That way we can delete the preStop hook.

For the servers though, the `preStop` hook doesn't work currently even without TLS. If I delete server pods, I'm seeing in the logs that it exits 1 instead of 0. I think it's because servers set `terminationGracePeriodSeconds` to 10s and so there's not enough time for it to properly exit. Unfortunately, leave_on_terminate has a hard-coded grace period of 15seconds which isn't enough for the servers. I'll keep the `preStop` hook there and increase `terminationGracePeriodSeconds` to 30sec (default).

ishustava commented on CHANGELOG.md:
```diff
@@ -11,6 +11,15 @@ IMPROVEMENTS:
 
+  * Optionally allow enabling TLS for servers and clients [[GH-313](https://github.com/hashicorp/consul-helm/pull/313/files#)].
+
+    Note that consul-k8s components don't currently work with HTTPS
```
I'll fix the injector before this PR gets merged.

ishustava commented on templates/tls-init-job.yaml:
```diff
@@ -0,0 +1,65 @@
+              consul tls ca create
+              consul tls cert create -server \
+                {{- range .Values.global.tls.serverAdditionalIPSANs }}
+                -additional-ipaddress={{ . }} \
```
Yeah I think that's a good idea.

ishustava commented on templates/tls-init-cleanup-job.yaml:
```diff
@@ -0,0 +1,48 @@
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  annotations:
+    "helm.sh/hook": pre-delete
```
No, because I want to make sure the resources it needs, like service account, cluster role etc, are still there. Post-delete executes after helm deletes all the resources.

ishustava commented on templates/tls-init-clusterrole.yaml:
```diff
@@ -0,0 +1,23 @@
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  annotations:
+    "helm.sh/hook": pre-install
```
You didn't need to because acl-init is a post-install,post-upgrade hook, and so it's executed after all resources are created. This is not the case for pre-install hooks. They are executed before any resources are created by Helm, so we need to mark these too to make sure they are there.

ishustava commented on templates/tls-init-clusterrole.yaml:
```diff
@@ -0,0 +1,23 @@
+    "helm.sh/hook-delete-policy": before-hook-creation
+rules:
+- apiGroups: [""]
+  resources:
```
Unfortunately, I can't. Kubernetes [doesn't support](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources) `resourceNames` for `create` requests.

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -235,6 +277,33 @@ spec:
+              -domain={{ .Values.global.domain }} \
+              -additional-ipaddress=${HOST_IP} \
+              -ca=/consul/tls/ca/tls.crt \
+              -key=/consul/tls/ca/tls.key
```
Hmm, that's an interesting question. For clients, these will be regenerated every time clients are restarted. So hopefully you restart your clients at least once in a year 🤞 

For servers, it might make sense to extend the validity period since those secrets are generated only once. We can extend those to two years.

Yeah, I could definitely rename the files :)

lkysow replied:
> For clients, these will be regenerated every time clients are restarted. So hopefully you restart your clients at least once in a year

Nice, nvm then.

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -235,6 +277,33 @@ spec:
+            consul tls cert create -client \
+              -dc={{ .Values.global.datacenter }} \
+              -domain={{ .Values.global.domain }} \
+              -additional-ipaddress=${HOST_IP} \
```
What is the use case for it? If we need it later for some reason, we could just add it to the cert, and it will get rotated on restart/upgrade. Eventually, I think we want to switch away from host IPs in favor of host-local traffic whenever kube supports it. In that case, we'd need to add a service name I think, but not sure what the actual implementation of it will be.

lkysow replied:
One use-case would be some other type of tooling. Mostly it's just for completeness.

ishustava replied:
I'm not sure this would be intended and have to think a bit more about security implications of this. I'll keep this out for now, but it should be easy to change if someone requests it.

ishustava commented on templates/server-statefulset.yaml:
```diff
@@ -99,6 +107,18 @@ spec:
                 -advertise="${POD_IP}" \
                 -bind=0.0.0.0 \
                 -bootstrap-expect={{ .Values.server.bootstrapExpect }} \
+                {{- if .Values.global.tls.enabled }}
```
I think we already discussed this for the clients. Will resolve this.

ishustava commented on templates/tls-init-cleanup-clusterrole.yaml:
```diff
@@ -0,0 +1,22 @@
+  resources:
+    - secrets
+  resourceNames:
+    - {{ template "consul.fullname" . }}-server-cert
```
It should be here! Will update

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -235,6 +277,33 @@ spec:
+            cd /consul/tls/client
+            consul tls cert create -client \
+              -dc={{ .Values.global.datacenter }} \
+              -domain={{ .Values.global.domain }} \
```
You're right they aren't strictly required, but if we don't add them, consul CLI assumes `dc1` and `.consul` domain and adds that as a common name for the client certificate. It doesn't look like it verifies it on RPC connections. This also might be useful if you are connecting to it from somewhere else and want to set that name as the SNI header.

ishustava commented:
@lkysow I've made the updates from your review and also added support for some consul-k8s components.
Changes made since your last review:
- Added TLS support for Connect inject (requires a consul-k8s docker image containing hashicorp/consul-k8s#181)
- Added TLS support to the Sync catalog deployment
- Added TLS support to the ACL init job ((requires a consul-k8s docker image containing hashicorp/consul-k8s#183)
- Now that consul-k8s supports TLS, switched `httpsOnly` to true by default. 

Outstanding work that will be submitted separately:
- TLS support for snapshot agents
- TLS support for mesh gateways

tpetrychyn commented:
Testing this branch in one of my clusters and I ran in to an issue that may be related to these changes.

Upgrading from consul-helm version 0.8.1 results in: 
```Error creating: pods "consul-consul-tls-init-" is forbidden: error looking up service account sd/consul-consul-tls-init: serviceaccount "consul-consul-tls-init" not found```

I opted to manually create a service account which allowed `consul-consul-tls-init` to report complete so I deleted it and re-ran helm upgrade but then consul-consul-server-2 would not initialize, giving the error: ```MountVolume.SetUp failed for volume "tls-server-cert" : secret "consul-consul-server-cert" not found```

I finally just blew away my existing consul installation and ran this chart clean.

I will say that running it on a clean environment worked great and the Golang Hashicorp Consul api with TlsConfig plugged in seamlessly so thank you for the work on this! Will be watching closely for using my own CA.

ishustava commented:
@tpetrychyn thank you so much for trying it out and leaving feedback here. Really appreciate it!

It looks like I forgot to mark the `tls-init-{service-account,clusterrole,clusterrolebinding}` as a `pre-upgrade` hook. I've made the change, and it should work now.

tpetrychyn commented:
Encountered another issue that I've been banging my head against for days..

Using commit 12ea51bb067d1c267d9f094d668f9c8eebea3d7e I am now getting the error in consul-sync-catalog:

```[GET /health/ready] Error getting leader status: Get https://10.137.132.250:8501/v1/status/leader: dial tcp 10.137.132.250:8501: i/o timeout```

I am using Digital Ocean managed k8 v1.16.2-do.1, my CNI is cilium v1.6.4. 

Using v0.8.1 of this chart works fine in this environment.

The oddest thing is 2 weeks I setup a completely clean DO cluster on 1.16.2-do.1 and this chart worked out of the box, but now even a clean cluster made today has that error..

P.S. commit ^6b659a2 breaks because consul-k8s 0.11.0 is not published on dockerhub

ishustava commented:
@tpetrychyn This PR requires a new version of consul-k8s for many components, such as Connect Inject and ACL bootstrapping, to work correctly. I optimistically updated the version. This PR, however, should not be merged without that docker image being published first. Sorry if that was confusing!

Have you tried deploying it with TLS disabled? Do you see the same error? At a first glance, it doesn't seem TLS related since the TLS handshake happens after the TCP connection is established.

tpetrychyn commented:
Thanks for the clarification! Non-tls also has that issue. Setting`exposeGossipPorts: true` appears to fix it.. Does that provide any insight?

ishustava commented on values.yaml:
```diff
@@ -77,6 +77,40 @@ global:
+    # Note: remember to switch it back to true once the rollout is complete.
+    # Please see this guide for more details:
+    # https://learn.hashicorp.com/consul/security-networking/certificates
+    verify: true
```
@lkysow this flag is for incremental TLS rollout

ishustava replied:
@adilyse ^^

lkysow commented on templates/client-daemonset.yaml:
```diff
@@ -166,17 +207,17 @@ spec:
             - name: aclconfig
               mountPath: /consul/aclconfig
             {{- end }}
-          lifecycle:
```
not something that needs to be changed but we might be adding this back as `consul force-leave -prune` due to https://github.com/hashicorp/consul/issues/6897#issuecomment-571153009

ishustava commented on CHANGELOG.md:
```diff
@@ -1,5 +1,15 @@
+    Note that this feature is only supported if both servers and clients are running
+    on Kubernetes. We will have better support for other deployment architectures,
+    as well as bringing your own CA, in the future.
+
```
done.

ishustava commented on templates/client-daemonset.yaml:
```diff
@@ -232,6 +280,38 @@ spec:
       {{- end }}
+      {{- if .Values.global.tls.enabled }}
+      - name: client-tls-init
+        image: {{ .Values.global.image }}
```
good point! thanks for calling this out. I'll update.

ishustava commented on templates/server-statefulset.yaml:
```diff
@@ -50,14 +50,22 @@ spec:
         {{ tpl .Values.server.tolerations . | nindent 8 | trim }}
     {{- end }}
-      terminationGracePeriodSeconds: 10
+      terminationGracePeriodSeconds: 30
```
Sometimes servers were taking longer than 10 seconds to greacefully leave the cluster. In those cases, kubernetes will just force kill the process, which in turn causes servers to exit 1 and not finish the graceful termination.

ishustava commented on templates/server-statefulset.yaml:
```diff
@@ -163,8 +205,14 @@ spec:
+                  {{- else }}
+                  curl http://127.0.0.1:8500/v1/status/leader \
+                  {{- end }}
+                  2>&1 | grep -E '".+"'
```
Good catch. I carried this from #137. Not sure why it was changed there, but I'll switch it back.

ishustava commented on templates/server-acl-init-job.yaml:
```diff
@@ -32,6 +32,12 @@ spec:
     spec:
       restartPolicy: Never
       serviceAccountName: {{ template "consul.fullname" . }}-server-acl-init
+      {{- if .Values.global.tls.enabled }}
```
Looking at the PSP for this job you've created in #326, I don't think anything will need to change once that's merged. This PR only adds secrets, and the PSP already allows attaching secret volumes. But let me know if I misunderstood you.

lkysow replied:
Yeah we're good now with my PR. That was just a reminder that we needed to do it for acl init.

ishustava commented on templates/tls-init-cleanup-job.yaml:
```diff
@@ -0,0 +1,52 @@
+# tls-init-cleanup job deletes Kubernetes secrets created by tls-init
```
done!

ishustava commented on templates/tls-init-job.yaml:
```diff
@@ -0,0 +1,79 @@
+                 {{- end }}
+                -dc={{ .Values.global.datacenter }} \
+                -domain={{ .Values.global.domain }}
+              curl -s -X POST --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
```
done.

ishustava commented:
@lkysow I've made the changes you requested.

fouadsemaan commented on templates/connect-inject-deployment.yaml:
```diff
@@ -60,19 +60,22 @@ spec:
                 -default-protocol="{{ .Values.connectInject.centralConfig.defaultProtocol }}" \
                 {{- end }}
-{{- if .Values.connectInject.certs.secretName }}
+                {{- if .Values.connectInject.certs.secretName }}
```
@ishustava does the generated envoy config file include the TLS context (cert/key)? I see that the CA is added inline.

ishustava replied:
The envoy TLS context for the consul client (aka `local_agent`) doesn't need cert/key since it's not using mTLS client auth to talk to Consul. That's why it only has the CA cert. 

If your question is about authentication, we recommend using ACLs for production environments for that purpose, and when they are enabled in the Helm chart, Envoy will include an ACL token when talking to Consul.

fouadsemaan replied:
Thank you. I was more curious about mTLS for service to service (via envoy) communication.

ishustava replied:
Ah, I see. mTLS for service to service is configured dynamically as you add and remove services by the Consul client agent. [These docs](https://www.consul.io/docs/connect/proxies/envoy#dynamic-configuration) talk a bit about that.

fouadsemaan replied:
Are you referring to the dynamically generated envoy-bootstrap.yaml file via the consul connect envoy -bootrap command? Is that where the cert/key should appear?

ishustava replied:
No, that config is used only for bootstrapping envoy itself. Consul updates envoy config for services dynamically through the API, so you won't find them in a file. You could look at the config though through the envoy admin endpoint. All you need to do is this:

```
kubectl port-forward po/<your pod that has envoy sidecar container> 19000
```

Then go to the browser at http://localhost:19000. There you will see an option to dump the config. Alternatively, you could just curl the endpoint to get the config `curl http://localhost:19000/config_dump`

You could do other things through that endpoint too, like change log levels. Here are the [docs](https://www.envoyproxy.io/docs/envoy/latest/operations/admin) for the admin endpoint in case you'll find them useful.

fouadsemaan replied:
Fantastic! I appreciate your help. Thank you! This is awesome!
